### PR TITLE
partial refactor

### PR DESCRIPTION
**Pull request by dfulu**
## Description

This is another part of the refactor.

- Move batch device handling and torch.no_grad into the forecast compiler
- Clean up the dataloaders and switching between the datapipes and data-sampler versions
- Clean up some of the logging set up
- Remove `PVNET_V2_VERSION` and `PVNET_V2_SUMMATION_VERSION` from docstrings as these aren't actually used anymore
- Reorder some of the `app()` function setup

### CODE REVIEW

File: src/pvnet_app/app.py [hunk 1]
```diff
@@ -126,61 +121,78 @@ def get_boolean_env_var(env_var: str, default: bool) -> bool:
...
+            value of -1, it will use one less than the number of CPU cores workers.
 
     This app expects these environmental variables to be available:
         - DB_URL
         - NWP_UKV_ZARR_PATH
         - NWP_ECMWF_ZARR_PATH
         - SATELLITE_ZARR_PATH
     The following are options
-        - PVNET_V2_VERSION, pvnet version, default is a version above
-        - PVNET_V2_SUMMATION_VERSION, the pvnet version, default is above
```
Comment (peterdudfield) [inline src/pvnet_app/app.py:143]: is possible could you update the readme on this project for the env vars. You might want to strip out the `environmental variables` text here as its then duplicate. Up to you

File: src/pvnet_app/app.py [hunk 2]
```diff
@@ -126,61 +121,78 @@ def get_boolean_env_var(env_var: str, default: bool) -> bool:
...
-    assert len(gsp_ids)>0, "No GSP IDs provided"
-
+    # Unpack the environment variables
     use_day_ahead_model = get_boolean_env_var("DAY_AHEAD_MODEL", default=False)
     use_ecmwf_only = get_boolean_env_var("USE_ECMWF_ONLY", default=False)
     run_extra_models = get_boolean_env_var("RUN_EXTRA_MODELS", default=False)
     use_ocf_data_sampler = get_boolean_env_var("USE_OCF_DATA_SAMPLER", default=True)
     use_adjuster = get_boolean_env_var("USE_ADJUSTER", default=True)
     save_gsp_sum = get_boolean_env_var("SAVE_GSP_SUM", default=False)
+    db_url = os.getenv("DB_URL")
```
Comment (peterdudfield) [inline src/pvnet_app/app.py:180]: cam we add something that errors if DB_URL is not set. There's a weird error message if we pass None into creating a connection string, that could be something we do in `DatabaseConnection` instead

File: src/pvnet_app/app.py [hunk 3]
```diff
@@ -336,8 +320,7 @@ def app(
     # ---------------------------------------------------------------------------
     # Escape clause for making predictions locally
     if not write_predictions:
-        temp_dir.cleanup()
-        return forecast_compilers[0].da_abs_all
+        return next(iter(forecast_compilers.values())).da_abs_all
```
Comment (peterdudfield) [inline src/pvnet_app/app.py:323]: this seems more complicated? Any reason?

File: src/pvnet_app/app.py [hunk 4]
```diff
@@ -285,46 +286,29 @@ def app(
...
 
     # ---------------------------------------------------------------------------
     # Make predictions
     logger.info("Processing batches")
 
-    s3_directory = os.getenv("SAVE_BATCHES_DIR", None)
+    for i, batch in enumerate(dataloader):
+        logger.info(f"Predicting for batch: {i}")
 
-    with torch.no_grad():
```
Comment (peterdudfield) [inline src/pvnet_app/app.py:315]: do you know what this did before?

**Review by peterdudfield — APPROVED** [review]
Ive left a few comments, and obviously only merge if the tests pass

File: src/pvnet_app/app.py (hunk 4 above)
Comment (dfulu) [reply src/pvnet_app/app.py:315]: Yeh, that made sure that ptorch is not storing gradients on the forward pass. I've just moved it inside the forecast compiler now. You'll see some of the methods decorated with `@torch.no_grad()`
